contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,189 @@
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev A complete binary tree with the ability to sequentially insert leaves, changing them from a zero to a non-zero
+ * value, while keeping a history of merkle roots. This structure allows inserting commitment (or other entries) that
+ * are not stored, but can be proven to be part of the tree.
+ *
+ * The history of merkle roots allow inclusion proofs to remain valid even if leaves are inserted into the tree between
+ *  the moment the proof is generated and the moment it's verified.
+ *
+ * Each tree can be customized to use specific
+ * - depth
+ * - length of the root history
+ * - zero values (for "empty" leaves)
+ * - hash function
+ *
+ * WARNING:
+ *
+ * By design, the tree include zero leaves. Customizing the "zero value" might be necessary to ensure that empty leaves
+ * being provably part of the tree is not a security issue.
+ *
+ * _Available since v4.x._
+ */
+library MerkleTree {
+    /**
+     * @dev Maximum supported depth. Beyond that, some checks will fail to properly work.
+     * This should be enough for any realistic usecase.
+     */
+    uint256 private constant MAX_DEPTH = 255;
+
+    /**
+     * @dev Merkle tree cannot be initialized because requested depth is to large.
+     */
+    error MerkleTreeInvalidDepth(uint256 depth, uint256 maxDepth);
+
+    /**
+     * @dev The `sides` and `zero` arrays are set, at initialization, to have a length equal to the depth of the tree.
+     * No push/pop operations should be performed of these array, and their lengths should not be updated.
+     *
+     * The `roots` array stores the history of roots. Its length is set at initialization, and should not be updated.
+     *
+     * The hashing function used during initialization to compute the `zeros` values (value of a node at a given depth
+     * for which the subtree is full of zero leaves). This function is kept in the structure for handling insertions.
+     */
+    struct TreeWithHistory {
+        uint256 currentRootIndex;
+        uint256 nextLeafIndex;
+        bytes32[] sides;
+        bytes32[] zeros;
+        bytes32[] roots;
+        function(bytes32, bytes32) view returns (bytes32) fnHash;
+    }
+
+    /**
+     * @dev Initialize using the default hash
+     */
+    function initialize(TreeWithHistory storage self, uint256 depth, uint256 length, bytes32 zero) internal {
+        return initialize(self, depth, length, zero, MerkleProof.hashPair);
+    }
+
+    /**
+     * @dev Initialize a new complete MerkleTree defined by:
+     * - Depth `depth`
+     * - All leaves are initialize to `zero`
+     * - Hashing function for a pair of leaves is fnHash
+     * and keep a root history of length `length` when leaves are inserted.
+     */
+    function initialize(
+        TreeWithHistory storage self,
+        uint256 depth,
+        uint256 length,
+        bytes32 zero,
+        function(bytes32, bytes32) view returns (bytes32) fnHash
```

Comment:
> How many fnHash variants people might end up? I personally think probably just sha256 and keccak256 so perhaps it's worth just an enum to avoid having function pointers.

They may want to use zk friendly hashes that we don't have implementation for.

Reply:
(example: tørnadø's MiMC)